libs/header/sync/sync.go

```diff
@@ -256,50 +246,83 @@ func (s *Syncer[H]) doSync(ctx context.Context, fromHead, toHead H) (err error)
 
 // processHeaders gets and stores headers starting at the given 'from' height up to 'to' height -
 // [from:to]
-func (s *Syncer[H]) processHeaders(ctx context.Context, from, to uint64) (int, error) {
-	headers, err := s.findHeaders(ctx, from, to)
-	if err != nil {
-		return 0, err
-	}
+// processHeaders checks headers in pending cache that apply to the requested range.
+// If some headers are missing, it starts requesting them from the network.
+// All headers that will be received are verified to be contiguous.
+func (s *Syncer[H]) processHeaders(
+	ctx context.Context,
+	fromHead H,
+	to uint64,
+) (err error) {
+	for {
+		headersRange, ok := s.pending.First()
+		if !ok {
+			break
+		}
 
-	amount, err := s.store.Append(ctx, headers...)
-	if err == nil && amount > 0 {
-		s.syncedHead.Store(&headers[amount-1])
-	}
-	return amount, err
-}
+		headers, amount := headersRange.Before(to)
+		if amount == 0 {
+			break
+		}
 
-// findHeaders gets headers from either remote peers or from local cache of headers received by
-// PubSub - [from:to]
-func (s *Syncer[H]) findHeaders(ctx context.Context, from, to uint64) ([]H, error) {
-	amount := to - from + 1 // + 1 to include 'to' height as well
-	if amount > s.Params.MaxRequestSize {
-		to = from + s.Params.MaxRequestSize - 1 // `from` is already included in range
-		amount = s.Params.MaxRequestSize
-	}
+		// check that returned range is adjacent to `fromHead`
+		if fromHead.Height()+1 != headers[0].Height() {
+			// make an external request
+			err = s.requestHeaders(ctx, fromHead, uint64(headers[0].Height()-1))
+			if err != nil {
+				return err
+			}
+		}
 
-	out := make([]H, 0, amount)
-	for from <= to {
-		// if we have some range cached - use it
-		r, ok := s.pending.FirstRangeWithin(from, to)
-		if !ok {
-			hs, err := s.exchange.GetRangeByHeight(ctx, from, to-from+1)
-			return append(out, hs...), err
+		// apply cached headers
+		if err = s.storeHeaders(ctx, headers); err != nil {
+			return err
 		}
+		// update fromHead for the next iteration
+		fromHead = headers[len(headers)-1]
+	}
+	return s.requestHeaders(ctx, fromHead, to)
+}
 
-		// first, request everything between from and start of the found range
-		hs, err := s.exchange.GetRangeByHeight(ctx, from, r.start-from)
+// requestHeaders requests headers from the network -> (fromHeader.Height;to].
+func (s *Syncer[H]) requestHeaders(
+	ctx context.Context,
+	fromHead H,
+	to uint64,
+) error {
+	amount := to - uint64(fromHead.Height())
+	// start requesting headers until amount will be 0
+	for amount > 0 {
+		size := s.Params.MaxRequestSize
+		if amount < size {
+			size = amount
+		}
+		headers, err := s.exchange.GetVerifiedRange(ctx, fromHead, size)
```

Comment:
Waiiit. I just realized. So GetVerifiedRange for exchange take an amount and store's one takes actual heigts :joy: We should fix this ASAP tbh

Reply:
Not necessary in this PR, but it's super cofooosing